### ISSUE TITLE
[language-toml] Whitespace is assumed to be invalid

### ISSUE THREAD

**Issue by savetheclocktower** [pulsar-edit/pulsar#564]
### Thanks in advance for your bug report!

- [X] Have you reproduced issue in [safe mode](https://pulsar-edit.dev/docs/atom-archive/faq/sections/what-does-safe-mode-do.html)?
- [X] Have you used the [debugging guide](https://pulsar-edit.dev/docs/atom-archive/hacking-atom/#debugging) to try to resolve the issue?
- [X] Have you checked our [FAQs](https://pulsar-edit.dev/docs/launch-manual/sections/faq/) to make sure your question isn't answered there?
- [X] Have you checked to make sure your issue does not already [exist](https://github.com/issues?q=is%3Aissue+user%3Apulsar-edit+is%3Aopen)?
- [X] Have you checked you are on the latest release of Pulsar?

### What happened?

As reported on DIscord, a TOML file that looks like this…

```toml
array = [
  0,
  1,
  2
]
```

…will highlight like this using the TextMate-style grammar:

<img width="83" alt="Screenshot 2023-05-29 at 12 27 03 PM" src="https://github.com/pulsar-edit/pulsar/assets/3450/6cf12851-6eb4-41c9-830b-25852179c8f1">

This is because it's got this pattern at the end of a list of `values` patterns in its grammar file, implying that anything not matched earlier in the list must therefore be invalid:

```js
      {
        'match': '.+'
        'name': 'invalid.illegal.toml'
      }
```

But the [TOML spec](https://toml.io/en/v1.0.0) does not come close to suggesting that whitespace is invalid in these situations.

(I've got a modern-tree-sitter version of the TOML grammar that fixes this, but it's not yet landed; it'll get added later on.)

### Pulsar version

1.105

### Which OS does this happen on?

❓ Other(Please specify in the OS details field below)

### OS details

(all OSes)

### Which CPU architecture are you running this on?

None

### What steps are needed to reproduce this?

Paste the given code sample into a TOML file.

**Comment by confused-Techie** [comment]
Resolved in TextMate grammars by #610 

Although, getting a Tree-Sitter grammar will be fantastic, and allow a much better implementation of this functionality.

**Comment by arite** [comment]
It appears that multi-line [TOML arrays](https://toml.io/en/v1.0.0#array) ([ABNF](https://github.com/toml-lang/toml/blob/1.0.0/toml.abnf#L197-L209)) unfortunately do not render correctly in Pulsar v1.107.1 so this issue may need re-opening.

For example the following TOML:
```toml
array1 = [ 0, 1, 2 ]

array2 = [
  0,
  1,
  2
]
```
Renders in Pulsar v1.107.1 as (using the TextMate grammar):

![image](https://github.com/pulsar-edit/pulsar/assets/2523293/7d9475e0-d196-4134-8d58-3682d291db21)

For comparison, using the modern Tree-Sitter implementation:

![image](https://github.com/pulsar-edit/pulsar/assets/2523293/12a22ba5-db0b-4cbd-a805-320a0f1ca99b)

--------------------

It looks like #610 addressed #574 (i.e. spaces in a single-line array), however when multiple lines are used the TextMate grammar continues to match on the `invalid.illegal.toml` rule on newline.

The above example is rendered correctly after making one of these changes:

* Adding an additional rule for `whitespace` to `values` which matches before `invalid.illegal.toml`
* Removing the `invalid.illegal.toml` rule `values` rule

This diff adds the `whitespace` rule to `values`:
```diff
diff --git a/packages/language-toml/grammars/toml.cson b/packages/language-toml/grammars/toml.cson
index 27d7b84e5..ece36299f 100644
--- a/packages/language-toml/grammars/toml.cson
+++ b/packages/language-toml/grammars/toml.cson
@@ -95,11 +95,11 @@
   'values':
     'patterns': [
       {
-        'begin': '\\[\\s*'
+        'begin': '\\['
         'beginCaptures':
           '0':
             'name': 'punctuation.definition.array.begin.toml'
-        'end': '\\s*\\]'
+        'end': '\\]'
         'endCaptures':
           '0':
             'name': 'punctuation.definition.array.end.toml'
@@ -108,7 +108,7 @@
             'include': '#comment'
           }
           {
-            'match': ',\\s*'
+            'match': ','
             'name': 'punctuation.definition.separator.comma.toml'
           }
           {
@@ -200,6 +200,9 @@
       {
         'include': '#numbers'
       }
+      {
+        'include': '#whitespace'
+      }
       {
         'match': '.+'
         'name': 'invalid.illegal.toml'
@@ -293,3 +296,6 @@
         'name': 'constant.numeric.binary.toml'
       }
     ]
+  'whitespace':
+    'match': '\\s+'
+    'name': 'punctuation.whitespace.ignore.toml'
```

Happy to submit a PR if the above solution seems OK, however there may be a better solution which only applies to e.g. the array rule.

**Comment by confused-Techie** [comment]
@arite Thanks a ton for pointing this out.

As our major focus will be on maintaining the Modern Tree-Sitter Grammars going forward, I think the solution of allowing any pieces of whitespace while maybe not always 100% syntactically correct, should be more than sufficient for the Text Mate Grammar.

So your solution adding a `whitespace` token to the `values` seems fantastic, and would encourage you to submit a PR for this solution if you'd like!
